File: interface/src/shared/api/memory.ts
import type {
  JsonValue,
  MemoryFact,
  MemoryEvent,
  MemoryProcedure,
  MemorySnapshot,
  MemoryStats,
  AgentContinuityConfig,
  MemoryContinuity,
  MemoryRetrievalTrace,
  MemoryAccessOptions,
} from "../types";
import { apiFetch } from "./core";

const memoryPath = (agentId: string, suffix = "") =>
  `/api/harness/agents/${encodeURIComponent(agentId)}/memory${suffix}`;

function withMemoryAccess(path: string, access?: MemoryAccessOptions, extra?: URLSearchParams) {
  const query = extra ?? new URLSearchParams();
  if (access?.projectId) query.set("project_id", access.projectId);
  if (access?.includeLegacy) query.set("include_legacy", "true");
  if (access?.scope) query.set("scope", access.scope);
  const encoded = query.toString();
  return encoded ? `${path}?${encoded}` : path;
}

export const memoryApi = {
  // Facts
  listFacts: (agentId: string, access?: MemoryAccessOptions) =>
    apiFetch<MemoryFact[]>(withMemoryAccess(memoryPath(agentId, "/facts"), access)),
  getFact: (agentId: string, factId: string, access?: MemoryAccessOptions) =>
    apiFetch<MemoryFact>(withMemoryAccess(memoryPath(agentId, `/facts/${factId}`), access)),
  getFactByKey: (agentId: string, key: string, access?: MemoryAccessOptions) =>
    apiFetch<MemoryFact>(withMemoryAccess(memoryPath(agentId, `/facts/by-key/${encodeURIComponent(key)}`), access)),
  createFact: (agentId: string, data: { key: string; value: JsonValue; confidence?: number; importance?: number; continuity?: MemoryContinuity }, access?: MemoryAccessOptions) =>
    apiFetch<MemoryFact>(withMemoryAccess(memoryPath(agentId, "/facts"), access), { method: "POST", body: JSON.stringify(data) }),
  updateFact: (agentId: string, factId: string, data: { key?: string; value?: JsonValue; confidence?: number; importance?: number; source?: MemoryFact["source"]; continuity?: MemoryContinuity }, access?: MemoryAccessOptions) =>
    apiFetch<MemoryFact>(withMemoryAccess(memoryPath(agentId, `/facts/${factId}`), access), { method: "PUT", body: JSON.stringify(data) }),
  deleteFact: (agentId: string, factId: string, access?: MemoryAccessOptions) =>
    apiFetch<void>(withMemoryAccess(memoryPath(agentId, `/facts/${factId}`), access), { method: "DELETE" }),

  // Events
  listEvents: (agentId: string, params?: { limit?: number; since?: string; event_type?: string }, access?: MemoryAccessOptions) => {
    const query = new URLSearchParams();
    if (params?.limit) query.set("limit", String(params.limit));
    if (params?.since) query.set("since", params.since);
    if (params?.event_type) query.set("event_type", params.event_type);
    return apiFetch<MemoryEvent[]>(withMemoryAccess(memoryPath(agentId, "/events"), access, query));
  },
  createEvent: (agentId: string, data: { event_type: string; summary: string; metadata?: JsonValue; importance?: number }, access?: MemoryAccessOptions) =>
    apiFetch<MemoryEvent>(withMemoryAccess(memoryPath(agentId, "/events"), access), { method: "POST", body: JSON.stringify(data) }),
  deleteEvent: (agentId: string, eventId: string, access?: MemoryAccessOptions) =>
    apiFetch<void>(withMemoryAccess(memoryPath(agentId, `/events/${eventId}`), access), { method: "DELETE" }),

  // Procedures
  listProcedures: (agentId: string, params?: { skill?: string; min_relevance?: number }, access?: MemoryAccessOptions) => {
    const query = new URLSearchParams();
    if (params?.skill) query.set("skill", params.skill);
    if (params?.min_relevance != null) query.set("min_relevance", String(params.min_relevance));
    return apiFetch<MemoryProcedure[]>(withMemoryAccess(memoryPath(agentId, "/procedures"), access, query));
  },
  createProcedure: (agentId: string, data: {
    name: string; trigger: string; steps: string[];
    context_constraints?: JsonValue; skill_name?: string; skill_relevance?: number;
    continuity?: MemoryContinuity;
  }, access?: MemoryAccessOptions) =>
    apiFetch<MemoryProcedure>(withMemoryAccess(memoryPath(agentId, "/procedures"), access), {
      method: "POST", body: JSON.stringify(data),
    }),
  updateProcedure: (agentId: string, procId: string, data: {
    name?: string; trigger?: string; steps?: string[];
    context_constraints?: JsonValue; skill_name?: string | null;
    skill_relevance?: number | null; success_rate?: number;
    continuity?: MemoryContinuity;
  }, access?: MemoryAccessOptions) =>
    apiFetch<MemoryProcedure>(withMemoryAccess(memoryPath(agentId, `/procedures/${procId}`), access), {
      method: "PUT", body: JSON.stringify(data),
    }),
  deleteProcedure: (agentId: string, procId: string, access?: MemoryAccessOptions) =>
    apiFetch<void>(withMemoryAccess(memoryPath(agentId, `/procedures/${procId}`), access), { method: "DELETE" }),

  // Aggregate
  getSnapshot: (agentId: string, access?: MemoryAccessOptions) =>
    apiFetch<MemorySnapshot>(withMemoryAccess(memoryPath(agentId), access)),
  getStats: (agentId: string, access?: MemoryAccessOptions) =>
    apiFetch<MemoryStats>(withMemoryAccess(memoryPath(agentId, "/stats"), access)),
  wipeMemory: (agentId: string, access?: MemoryAccessOptions) =>
    apiFetch<void>(withMemoryAccess(memoryPath(agentId), access), { method: "DELETE" }),
  triggerConsolidation: (agentId: string, access?: MemoryAccessOptions) =>
    apiFetch<void>(withMemoryAccess(memoryPath(agentId, "/consolidate"), access), { method: "POST" }),
  getContinuityConfig: (agentId: string) =>
    apiFetch<AgentContinuityConfig>(memoryPath(agentId, "/continuity")),
  updateContinuityConfig: (agentId: string, config: AgentContinuityConfig) =>
    apiFetch<AgentContinuityConfig>(memoryPath(agentId, "/continuity"), {
      method: "PUT",
      body: JSON.stringify(config),
    }),
  getLatestRetrievalTrace: (agentId: string, access?: MemoryAccessOptions) =>
    apiFetch<MemoryRetrievalTrace | null>(withMemoryAccess(memoryPath(agentId, "/retrieval/latest"), access)),
};

File: apps/aura-os-server/src/handlers/harness_proxy/memory.rs
use axum::extract::{Path, RawQuery, State};
use axum::http::{Method, StatusCode};
use axum::response::Response;

use aura_os_core::{AgentId, ZeroAuthSession};

use super::require_agent_proxy_access;
use crate::state::{AppState, AuthJwt, AuthSession};

async fn proxy_agent_json(
    state: &AppState,
    jwt: &str,
    session: &ZeroAuthSession,
    agent_id: &AgentId,
    method: Method,
    path: String,
    query: Option<String>,
    body: Option<String>,
) -> Result<Response, StatusCode> {
    require_agent_proxy_access(state, jwt, session, agent_id).await?;
    let query = authenticated_memory_query(state, jwt, session, agent_id, query).await?;
    state
        .harness_http
        .proxy_json(method, &path, query, body)
        .await
}

async fn authenticated_memory_query(
    state: &AppState,
    jwt: &str,
    session: &ZeroAuthSession,
    agent_id: &AgentId,
    query: Option<String>,
) -> Result<Option<String>, StatusCode> {
    let (mut pairs, project_id) = canonical_memory_query(query.as_deref(), &session.user_id)?;

    if let Some(project_id) = project_id.as_deref() {
        let storage = state
            .storage_client
            .as_deref()
            .ok_or(StatusCode::SERVICE_UNAVAILABLE)?;
        let bindings = crate::handlers::agents::chat::find_matching_project_agents(
            state,
            storage,
            jwt,
            &agent_id.to_string(),
        )
        .await;
        let has_binding = bindings
            .iter()
            .any(|binding| binding.project_id.as_deref() == Some(project_id));
        if !has_binding
            && !owned_session_matches_project(storage, jwt, agent_id, project_id).await?
        {
            return Err(StatusCode::FORBIDDEN);
        }
    }

    if let Some(project_id) = project_id {
        pairs.push(("project_id".to_string(), project_id));
    }
    let encoded = url::form_urlencoded::Serializer::new(String::new())
        .extend_pairs(pairs)
        .finish();
    Ok(Some(encoded))
}

async fn owned_session_matches_project(
    storage: &aura_os_storage::StorageClient,
    jwt: &str,
    agent_id: &AgentId,
    project_id: &str,
) -> Result<bool, StatusCode> {
    let agent_id = agent_id.to_string();
    let sessions = storage
        .list_my_sessions(jwt)
        .await
        .map_err(|_| StatusCode::SERVICE_UNAVAILABLE)?;
    Ok(sessions.iter().any(|entry| {
        entry.agent_id.as_deref() == Some(agent_id.as_str())
            && entry.session.project_id.as_deref() == Some(project_id)
    }))
}

type MemoryQueryPairs = Vec<(String, String)>;

fn canonical_memory_query(
    query: Option<&str>,
    authenticated_user_id: &str,
) -> Result<(MemoryQueryPairs, Option<String>), StatusCode> {
    let mut pairs: MemoryQueryPairs = query
        .map(|query| {
            url::form_urlencoded::parse(query.as_bytes())
                .filter(|(key, _)| key != "user_id")
                .map(|(key, value)| (key.into_owned(), value.into_owned()))
                .collect()
        })
        .unwrap_or_default();

    let project_ids: std::collections::HashSet<String> = pairs
        .iter()
        .filter(|(key, value)| key == "project_id" && !value.trim().is_empty())
        .map(|(_, value)| value.clone())
        .collect();
    if project_ids.len() > 1 {
        return Err(StatusCode::BAD_REQUEST);
    }
    let project_id = project_ids.into_iter().next();
    pairs.retain(|(key, _)| key != "project_id");

    pairs.push(("user_id".to_string(), authenticated_user_id.to_string()));
    Ok((pairs, project_id))
}

pub(crate) async fn list_facts(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/facts"),
        query,
        None,
    )
    .await
}

pub(crate) async fn get_fact(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, fact_id)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/facts/{fact_id}"),
        query,
        None,
    )
    .await
}

pub(crate) async fn create_fact(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
    body: String,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::POST,
        format!("api/agents/{agent_id}/memory/facts"),
        query,
        Some(body),
    )
    .await
}

pub(crate) async fn update_fact(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, fact_id)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
    body: String,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::PUT,
        format!("api/agents/{agent_id}/memory/facts/{fact_id}"),
        query,
        Some(body),
    )
    .await
}

pub(crate) async fn delete_fact(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, fact_id)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::DELETE,
        format!("api/agents/{agent_id}/memory/facts/{fact_id}"),
        query,
        None,
    )
    .await
}

pub(crate) async fn get_fact_by_key(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, key)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/facts/by-key/{key}"),
        query,
        None,
    )
    .await
}

pub(crate) async fn list_events(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/events"),
        query,
        None,
    )
    .await
}

pub(crate) async fn create_event(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
    body: String,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::POST,
        format!("api/agents/{agent_id}/memory/events"),
        query,
        Some(body),
    )
    .await
}

pub(crate) async fn delete_event(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, event_id)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::DELETE,
        format!("api/agents/{agent_id}/memory/events/{event_id}"),
        query,
        None,
    )
    .await
}

pub(crate) async fn list_procedures(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/procedures"),
        query,
        None,
    )
    .await
}

pub(crate) async fn get_procedure(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, proc_id)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/procedures/{proc_id}"),
        query,
        None,
    )
    .await
}

pub(crate) async fn create_procedure(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
    body: String,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::POST,
        format!("api/agents/{agent_id}/memory/procedures"),
        query,
        Some(body),
    )
    .await
}

pub(crate) async fn update_procedure(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, proc_id)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
    body: String,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::PUT,
        format!("api/agents/{agent_id}/memory/procedures/{proc_id}"),
        query,
        Some(body),
    )
    .await
}

pub(crate) async fn delete_procedure(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, proc_id)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::DELETE,
        format!("api/agents/{agent_id}/memory/procedures/{proc_id}"),
        query,
        None,
    )
    .await
}

pub(crate) async fn list_procedures_by_skill(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path((agent_id, skill_name)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    let mut qs = format!("skill={skill_name}");
    if let Some(q) = query {
        qs = format!("{qs}&{q}");
    }
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/procedures"),
        Some(qs),
        None,
    )
    .await
}

pub(crate) async fn get_memory_snapshot(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory"),
        query,
        None,
    )
    .await
}

pub(crate) async fn wipe_memory(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::DELETE,
        format!("api/agents/{agent_id}/memory"),
        query,
        None,
    )
    .await
}

pub(crate) async fn get_memory_stats(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/stats"),
        query,
        None,
    )
    .await
}

pub(crate) async fn trigger_consolidation(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
    body: String,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::POST,
        format!("api/agents/{agent_id}/memory/consolidate"),
        query,
        Some(body),
    )
    .await
}

pub(crate) async fn get_continuity_config(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/continuity"),
        None,
        None,
    )
    .await
}

pub(crate) async fn update_continuity_config(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    body: String,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::PUT,
        format!("api/agents/{agent_id}/memory/continuity"),
        None,
        Some(body),
    )
    .await
}

pub(crate) async fn get_latest_retrieval_trace(
    State(state): State<AppState>,
    AuthJwt(jwt): AuthJwt,
    AuthSession(session): AuthSession,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, StatusCode> {
    proxy_agent_json(
        &state,
        &jwt,
        &session,
        &agent_id,
        Method::GET,
        format!("api/agents/{agent_id}/memory/retrieval/latest"),
        query,
        None,
    )
    .await
}

#[cfg(test)]
mod tests {
    use super::canonical_memory_query;
    use axum::http::StatusCode;

    #[test]
    fn memory_query_replaces_spoofed_user_identity() {
        let (pairs, project_id) = canonical_memory_query(
            Some("project_id=project-a&user_id=attacker&include_legacy=true"),
            "signed-in-user",
        )
        .expect("query should canonicalize");

        assert_eq!(project_id.as_deref(), Some("project-a"));
        assert_eq!(
            pairs
                .iter()
                .filter(|(key, _)| key == "user_id")
                .map(|(_, value)| value.as_str())
                .collect::<Vec<_>>(),
            vec!["signed-in-user"]
        );
        assert!(pairs
            .iter()
            .any(|(key, value)| key == "include_legacy" && value == "true"));
    }

    #[test]
    fn memory_query_rejects_ambiguous_project_identity() {
        let error = canonical_memory_query(
            Some("project_id=project-a&project_id=project-b"),
            "signed-in-user",
        )
        .expect_err("two project identities must fail closed");
        assert_eq!(error, StatusCode::BAD_REQUEST);
    }
}
